source/MaterialXGenOgsXml/GlslFragmentGenerator.cpp

```diff
@@ -0,0 +1,192 @@
+
+    // Turn on fixed float formatting to make sure float values are
+    // emitted with a decimal point and not as integers, and to avoid
+    // any scientific notation which isn't supported by all OpenGL targets.
+    Value::ScopedFloatFormatting fmt(Value::FloatFormatFixed);
+
+    const ShaderGraph& graph = shader->getGraph();
+
+    // Add global constants and type definitions
+    emitInclude("pbrlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_defines.glsl", context, stage);
+    const unsigned int maxLights = std::max(1u, context.getOptions().hwMaxActiveLightSources);
+    emitLine("#define MAX_LIGHT_SOURCES " + std::to_string(maxLights), stage, false);
+    emitLineBreak(stage);
+    emitTypeDefinitions(context, stage);
+
+    // Add all constants
+    const VariableBlock& constants = stage.getConstantBlock();
+    if (!constants.empty())
+    {
+        emitVariableDeclarations(constants, _syntax->getConstantQualifier(), SEMICOLON, context, stage);
+        emitLineBreak(stage);
+    }
+
+    bool lighting = graph.hasClassification(ShaderNode::Classification::SHADER | ShaderNode::Classification::SURFACE) ||
+        graph.hasClassification(ShaderNode::Classification::BSDF);
+
+    // Emit common math functions
+    emitInclude("pbrlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_math.glsl", context, stage);
+    emitLineBreak(stage);
+
+    // Emit lighting functions
+    if (lighting)
+    {
+        if (context.getOptions().hwSpecularEnvironmentMethod == SPECULAR_ENVIRONMENT_FIS)
+        {
+            emitInclude("pbrlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_environment_fis.glsl", context, stage);
+        }
+        else
+        {
+            emitInclude("pbrlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_environment_prefilter.glsl", context, stage);
+        }
+        emitLineBreak(stage);
+    }
+
+    // Emit sampling code if needed
+    if (graph.hasClassification(ShaderNode::Classification::CONVOLUTION2D))
+    {
+        // Emit sampling functions
+        emitInclude("stdlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_sampling.glsl", context, stage);
+        emitLineBreak(stage);
+    }
+
+    // Emit uv transform function
+    if (context.getOptions().fileTextureVerticalFlip)
+    {
+        emitInclude("stdlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_get_target_uv_vflip.glsl", context, stage);
+        emitLineBreak(stage);
+    }
+    else
+    {
+        emitInclude("stdlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_get_target_uv_noop.glsl", context, stage);
+        emitLineBreak(stage);
+    }
+
+    // Add all functions for node implementations
+    emitFunctionDefinitions(graph, context, stage);
+
+    const ShaderGraphOutputSocket* outputSocket = graph.getOutputSocket();
+
+    // Add function signature
+    string functionName = shader->getName();
+    _syntax->makeValidName(functionName);
+    emitLine("vec4 " + functionName, stage, false); // TODO: We are always outputting vec4 for now
```

Comment:
Not sure if this is done elsewhere but we need the function signature (name) to be accessible.

Reply:
Also guess the vec4 needs to be change to vec3.

Reply:
Ah yes, will fix this!